trace/es/elasticsearch.go

```
@@ -0,0 +1,185 @@
+	DefaultPort = "9200"
+
+	// span tag that denotes the host that responded to a query
+	respondentTag = "respondent"
+)
+
+type tracingInfo struct {
+	user  string
+	hosts []string
+}
+
+func (t *tracingInfo) startSpan(opName string, req *http.Request) (opentracing.Span, error) {
+
+	if req == nil {
+		return nil, fmt.Errorf("request is empty")
+	}
+
+	uri := req.URL.RequestURI()
+	method := req.Method
+
+	var bodyFmt string
+	if req.Body != nil {
+		if rawBody, err := ioutil.ReadAll(req.Body); err == nil {
+			req.Body = ioutil.NopCloser(bytes.NewReader(rawBody))
+			bodyFmt = string(rawBody)
+		}
+	}
+
+	return trace.EsSpan(req.Context(), opName, "go-elasticsearch", t.user, uri, method, bodyFmt, t.hosts), nil
+}
+
+func endSpan(sp opentracing.Span, rsp *http.Response) {
+	// Fetch the host that responded and inject it to the span
+	sp.SetTag(respondentTag, rsp.Request.URL.Host)
+
+	trace.FinishHTTPSpan(sp, rsp.StatusCode)
+}
+
+// transportClient is ES transport struct wrapped with tracing info
+type transportClient struct {
+	client *estransport.Client
+	tracingInfo
+}
+
+func (c *transportClient) Perform(req *http.Request) (*http.Response, error) {
+	// First try to retrieve the name of the caller to pass opName
+	var opName string
+	pc, _, _, ok := runtime.Caller(1)
+	if ok {
+		if opInfo := runtime.FuncForPC(pc); opInfo != nil {
+			opName = opInfo.Name()
+		}
+	} else {
+		opName = "Transport.Perform"
+	}
+
+	sp, err := c.startSpan(opName, req)
+	if err != nil {
+		return nil, err
+	}
+
+	rsp, err := c.client.Perform(req)
+	if err != nil || rsp == nil {
+		trace.SpanError(sp)
+		return rsp, err
+	}
+	endSpan(sp, rsp)
+	return rsp, nil
+}
+
+// Config is a wrapper for elasticsearch.Config
+type Config elasticsearch.Config
+
+// Client is a wrapper for elasticsearch.Client
+type Client struct {
+	elasticsearch.Client
+}
+
+// NewDefaultClient returns an empty ES client with sane defaults
+func NewDefaultClient() *Client {
+	client, _ := NewClient(Config{})
```

error is not checked.